pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -773,10 +775,10 @@ func implicitWaypointPolicies(ctx krt.HandlerContext, Waypoints krt.Collection[W
 		return nil
 	}
 	serviceWaypointKeys := slices.MapFilter(services, func(si model.ServiceInfo) *string {
-		if si.Waypoint == "" || (waypoint != nil && waypoint.ResourceName() == si.Waypoint) {
+		if si.Waypoint.ResourceName == "" || (waypoint != nil && waypoint.ResourceName() == si.Waypoint.ResourceName) {
```

Comment:
**refactor nit:**

Do you think it makes sense check if waypoint is nil (the one passed into `implicitWaypointPolicies`) instead of checking if waypoint is nil multiple times throughout the function? Or does that break intended functionality? It seems like when we check if it's nil, we just return nil. We can do that once in the beginning and be done with doing the check. That particular Waypoint value won't change throughout the code as far as I can tell, right? Then we'd be able to simplify the logic in this function a little bit, remove some redundant lines of code.

instead:
```go
func implicitWaypointPolicies(ctx krt.HandlerContext, Waypoints krt.Collection[Waypoint], waypoint *Waypoint, services []model.ServiceInfo) []string {
	if !features.DefaultAllowFromWaypoint || waypoint == nil {
		return nil
	}
	serviceWaypointKeys := slices.MapFilter(services, func(si model.ServiceInfo) *string {
		if si.Waypoint.ResourceName == "" || waypoint.ResourceName() == si.Waypoint.ResourceName {
			return nil
		}
		return ptr.Of(si.Waypoint.ResourceName)
	})
	if len(serviceWaypointKeys) == 0 {
		n := implicitWaypointPolicyName(waypoint)
		if n != "" {
			return []string{waypoint.Namespace + "/" + n}
		}
	}
	waypoints := krt.Fetch(ctx, Waypoints, krt.FilterKeys(serviceWaypointKeys...))
	waypoints = append(waypoints, *waypoint)

	return slices.MapFilter(waypoints, func(w Waypoint) *string {
		policy := implicitWaypointPolicyName(&w)
		if policy == "" {
			return nil
		}
		return ptr.Of(w.Namespace + "/" + policy)
	})
}
```

Reply:
I think you can get to the `return` with waypoint==nil currently. if si.Waypoint.ResourceName!="" then serviceWaypointKeys will exist which could get a `waypoints`... but TBH I am not really sure, it seems odd we are looking up service waypoints when looking at waypoints attach to a pods..

Reply:
> it seems odd we are looking up service waypoints when looking at waypoints attach to a pods..

That's a good point.. we might want to confirm quickly if/why that is needed. Definitely seems like a code smell now that you've highlighted it.